Replication rework

Vagabond opened the pull request:
This major rework of the replication code provides the following benefits:
- Per bucket replication
- Generic hooks for per-object replication decisions (eg search ensuring the proxy object is replicated along with the KV object).
- Split fullsync and realtime replication into different modules, so fullsync can be pluggable.
- Implement a new 'keylist' fullsync strategy.
- Make the TCP send/recv buffers tunable.

Please treat this as a full code review, don't just look at the changes, because these are very extensive modifications.

Vagabond committed:
Split the fullsync repl from the realtime repl

This allows pluggable fullsync strategies, so we can experiment with
better ones.

Vagabond committed:
Support configuring TCP send/receive buffer sizes

Vagabond committed:
Re-add code to handle tcp_closed/tcp_error

Vagabond committed:
Make repl helpers pluggable, and move the search-specific code into search

Vagabond committed:
Implement per-bucket repl; don't needlessly create new clients

Vagabond committed:
Add a realtime repl hook for applications to modify realtime repl

Vagabond committed:
Some reorgs to make writing a new fullsync strategy easier

Vagabond committed:
Initial implementation of new fullsync strategy

This strategy has the client send its keylist to the server, which diffs
it with its own keylist and sends the differences. No merkle trees are
generated and I can mostly complete a fullsync with 20M keys.

Vagabond committed:
Remember replication progress, and print # of remaining partitions

Vagabond committed:
Use a worker pool for repl puts, not unlimited spawned processes

Vagabond committed:
Cleanup some obsolete code

Vagabond committed:
Workaround for the stale poolboy we're using

Vagabond committed:
Logging for each phase of replication, with elapsed time

Vagabond committed:
Implement start/cancel/pause/resume for keylist fullsync

Vagabond committed:
Fix the fullsync stats, more log messages

Vagabond committed:
Write into the keylist directly from the fold, fix GC runs for filesorter

Vagabond committed:
Skip partitions missing on the client or the server side

Vagabond committed:
Fix a couple minor bugs

Vagabond committed:
Fix start/stop/pause/resume for legacy strategy

Vagabond committed:
Fixes to schedule_fullsync and avoid double-skipping a partition

Double skipping means when the partition is not available on either
side.

Vagabond committed:
Reorder receive statement so stops are checked for first

Also, cleanup the client's keylist after its been sent.

Vagabond committed:
Some final cleanups

Vagabond committed:
Re-add repl status command support

beerriot commented:
Something that would help reviewing, and I think would also be a big improvement in its own right, would be an English readme describing how this version of replication works.  The `replication.org` file in `riak_ee/doc` gives something of a high-level explanation, but what I'm hoping for is something a little bit more specific about the guts: what each module does, where apps hook into repl, details of the inter-cluster protocol, etc.  This can all be learned by reading the code (though more doc'ing there could also help), but I think an overall doc could help both in learning, and also in spotting inconsistencies between how we think it works, and how it actually does.

Vagabond commented:
Yes, this is actually something I meant to do, but I forgot. Will type something up.

Vagabond commented:
Added documentation, and opened a new P/R on riak_ee for the details on the replication hooks in the replication.org file.

Vagabond committed:
Add a bunch of documentation about the fullsync protocols

Vagabond committed:
Fix another process leak; add put pool size to client stats

The put_fsms in the repl_do_put function were not blocking, so while the
worker pool was managing the # of workers, the # of put processes was
not being controlled. Now, the worker waits for the put_fsm to reply
before it returns and handles the next piece of work.

Vagabond committed:
Fix long-broken leader stats

Vagabond committed:
3 strategies to help avoid TCP logjams

* Periodically ACK the keylist chunks, to provide backpressure.
* Differentiate realtime and fullsync diff_objs, so that fullsync
  object exchange doesn't count towards realtime objects received.
* Set the socket to close after a 5 second send timeout, to avoid worst
  case where repl will stall forever.

Vagabond committed:
Re-implement site redirection

Vagabond committed:
Avoid message send timeouts under load

beerriot commented on src/riak_repl_tcp_server.erl:
```diff
@@ -147,373 +189,61 @@ wait_peerinfo({peerinfo, TheirPeerInfo, Capability},
+    case riak_repl_leader:leader_node()  of
+        undefined -> % leader not elected yet
+            %% check again in 5 seconds
+            erlang:send_after(5000, self(), election_wait),
```
It looks like the code to handle this message has been removed.  Grepping for `election_wait` finds only this line.

Vagabond committed:
Re-add handler for election_wait messages

beerriot commented on src/riak_repl_tcp_server.erl:
```diff
@@ -147,373 +191,61 @@ wait_peerinfo({peerinfo, TheirPeerInfo, Capability},
+            send(Socket, {peerinfo, PI,
+                [bounded_queue, {fullsync_strategies,
+                        app_helper:get_env(riak_repl, fullsync_strategies,
+                            [?LEGACY_STRATEGY])}]}),
```
This call might return `{error, not_leader}` if the leader has changed since the call to `riak_repl_leader:leader_node/0`.

beerriot commented on src/riak_repl_tcp_server.erl:
```diff
@@ -147,373 +191,61 @@ wait_peerinfo({peerinfo, TheirPeerInfo, Capability},
+        OurNode ->
+            erlang:cancel_timer(State#state.election_timeout),
+            %% this switches the socket into active mode
+            Props = riak_repl_fsm_common:common_init(Socket),
```
The return value of `send/2` is ignored here, but could be either `ok` or `{error, Reason}`.

Vagabond committed:
Handle change of leaders during peerinfo sending

beerriot commented on src/riak_repl_fullsync_helper.erl:
```diff
@@ -125,13 +149,16 @@ handle_call({make_keylist, Partition, Filename}, From, State) ->
                              %% accept there could be a potentially huge message queue
                              Folder = fun(K, V, MPid) -> 
-                                              gen_server2:cast(MPid, {kl, K, hash_object(V)}),
+                                              %gen_server2:cast(MPid, {kl, K, hash_object(V)}),
```
This code should be removed if unused.  Also, it seems to be the only place casting the `{kl, K, H}` tuple, so if it's truly unused, the handle_cast clause for this tuple should be removed as well.

beerriot commented on src/riak_repl_fullsync_helper.erl:
```diff
@@ -321,10 +354,18 @@ open_couchdb(Filename) ->
     {ok, Fd, Btree}.
 
 itr_new(File, Tag) ->
+    case itr_fresh(File, Tag) of
```
It looks like this is the only call to `itr_fresh/2`. Why add the extra fun wrapper if it's just going to be evaluated immediately like it used to be?

beerriot commented:
Further comments on `riak_repl_fullsync_helper`, from full review, not related to changes in this PR:
- The process is `gen_server2` behavior, but `gen_server` calls are made in a few places
- The `{make_keylist, Partition, Filename}` clause of `handle_call` does uses both a `gen_server:reply` and a `{reply, Reply, State}` tuple, which seems like potential for trouble. (Why did I also think it was necessary to return a `noreply` result _before_ calling `gen_server:reply`?)
- The use of R/remote and L/local in `diff_hash/2` is confusing when used in the new `keylist` strategy, since the code is running on the "server" side of repl, but the Remote argument is the server's keylist.

Vagabond committed:
Cleanup dead code in the fullsync_helper

Vagabond committed:
Don't double-reply, use gen_server2 calls from a gen_server2

beerriot commented on src/riak_repl_util.erl:
```diff
@@ -44,48 +52,33 @@ get_partitions(_Ring) ->
+            K = riak_object:key(Object),
+            Opts = [asis, disable_hooks, {update_last_modified, false}],
+
+            riak_kv_put_fsm_sup:start_put_fsm(node(), [ReqId, Object, 1, 1,
```
Should this link to the Pid that is returned, so the worker tears down if the put fsm barfs?  I guess the timeout in the receive will trigger otherwise…

beerriot commented on src/riak_repl_util.erl:
```diff
@@ -93,6 +86,111 @@ reap(ReqId, B, K) ->
+    receive
+        {ReqId, _} ->
+            ok
+    after 60000 ->
```
Why 60sec instead of `?REPL_FSM_TIMEOUT`?

beerriot commented on src/riak_repl_util.erl:
```diff
@@ -93,6 +86,111 @@ reap(ReqId, B, K) ->
 
+wait_for_response(ReqId) ->
+    receive
+        {ReqId, _} ->
```
Should the reply be checked, with maybe a `lager:debug` if the put failed?

Vagabond committed:
Debug logging if putting/reaping a repled object fails

beerriot committed:
doc typos

beerriot commented on src/riak_repl_tcp_server.erl:
```diff
@@ -1,144 +1,188 @@
+        Objects when is_list(Objects) ->
+            %% enqueue all the objects the hook asked us to send
+            NewState = lists:foldl(fun(O, S) ->
+                        enqueue(term_to_binary({diff_obj, O}), S)
```
This seems like it's likely to break the guarantees made to repl helpers/hooks about replicating some objects before others.  The `bounded_queue` module dequeues items to make space for new items when its limit is reached.  So, if the repl helper/hook returned a sufficiently large list of objects, or a list of sufficiently-large-size objects, and the repl link was unable to keep up, the objects early in the in that list would be dropped on the floor, while the objects later in the list would be replicated.

That is:

```
22> Q0 = bounded_queue:new(3).       
{bq,3,0,{[],[]},0}
23> Q1 = bounded_queue:in(Q0, <<1>>).
{bq,3,1,{[<<1>>],[]},0}
24> Q2 = bounded_queue:in(Q1, <<2>>).
{bq,3,2,{[<<2>>],[<<1>>]},0}
25> Q3 = bounded_queue:in(Q2, <<3>>).
{bq,3,3,{[<<3>>,<<2>>],[<<1>>]},0}
% Q3 is at capacity

26> bounded_queue:in(Q3, <<4>>). % add one more small entry
{bq,3,3,{[<<4>>,<<3>>],[<<2>>]},1}
% The first entry, <<1>>, has been lost

27> bounded_queue:in(Q3, <<5,6,7,8>>). % add one larger-than-capacity entry
{bq,3,4,{[<<5,6,7,8>>],[]},3}
% All previous entries have been lost
```

Vagabond replied:
This is an excellent point, do you have any suggestions for solving it? Maybe just sending those objects directly and only enqueueing the original object?

beerriot replied:
The only functional workaround I can think of is to add all of the "dependencies" and the root object as one, combined element in the queue.  The combined element may be large enough to kick other things out, but they would either all make it to send time, or none of them would.

The nonfunctional "workaround" was to change `bounded_queue` so that it refuses to enqueue new items, instead of dropping old ones, when its threshold is reached.  It prevents replicating an object before its dependencies, but may also prevent _ever_ replicating an object (because its dependencies fill the queue).

Vagabond replied:
Yes, I had the same solution come to me in a dream.

Vagabond replied:
Alright, I pushed a patch implementing this, with a test (and some manual confirmation it works).

Vagabond committed:
Also allow lists of objects to be placed in the bounded_queue

beerriot commented:
+1 merge (see AZ ticket for testing details/omissions)

jonmeredith commented on src/bounded_queue.erl:
```diff
@@ -43,31 +43,38 @@
         false ->
             make_fit(BQ, Item, ItemSize)
+    end;
+in(BQ=#bq{m=Max, q=Q, d=D}, [H|_T] = Items) when is_binary(H) ->
```
Could do with a comment to explain the list is added as an item in the queue, rather than adding a list of items to the queue individually.
